File: solutions_5648941810974720_1/C++/sergio10barca/Getting the Digits.cpp
#include <iostream>
#include <string>
#include <map>
#include <algorithm>
#include <cstdio>
using namespace std;

bool buscar(int vec[], string s){
//    cout << "buscar " << s << endl;
    int ant[200];
    for(int i = 0; i < 200; i++)
        ant[i] = vec[i];
    for(int i = 0; i < s.size(); i++){
        if(ant[s[i]] <= 0){
            return false;
        }
        else
            ant[s[i]]--;
    }
    for(int i = 0; i < s.size(); i++){
        vec[s[i]]--;
    }
    return true;
}

string solve(string s){
    string v[] = {"ZERO", "TWO", "SIX", "EIGHT", "THREE", "FOUR",
    "FIVE", "SEVEN", "ONE", "NINE"};
    /*(Z, 0), (2, W), (6, X), (8, G)
    (H, 3), (4, R), (5, F), (7, V), (o, 1), (N, 9)*/
    string t = "";
    int vec[200] = {0};
    for(int i = 0; i < s.size(); i++){
        vec[s[i]]++;
    }
    int resp[] = {0, 2, 6, 8, 3, 4, 5, 7, 1, 9};
    for(int i = 0; i < 10; i++){
        while(buscar(vec, v[i])){
            t+=char(resp[i]+'0');
        }
    }
    for(int i = 0; i < 200; i++)
        if(vec[i] > 0)
            while(true);
    sort(t.begin(), t.end());
    return t;
}

int main(){
    int casos;
    string s;
    cin >> casos;
    for(int i = 1; i <= casos; i++){
        cin >> s;
        printf("Case #%d: %s\n", i, solve(s).c_str());
    }
    return 0;
}

File: solutions_5648941810974720_0/Python/martinr/Getting the Digits.py
 
def solve(s):
    def fits(digit, cnt):
        s = set(digit)
        for c in s:
            if not(c in cnt) or (digit.count(c) > cnt[c]):
                return False
        return True
    
    A = ['SIX', 'EIGHT', 'TWO', 'ZERO', 'FIVE',  'THREE', 'FOUR', 'NINE', 'SEVEN', 'ONE']

    D = {'ZERO':'0','ONE':'1', 'TWO':'2', 'THREE':'3','FOUR':'4', 'FIVE':'5', 'SIX':'6', 'SEVEN':'7', 'EIGHT':'8', 'NINE':'9'}
    cnt={}
    for c in set(s):
        cnt[c] = s.count(c)
        
    l = 0
    res = []
    for d in A:         
        while fits(d, cnt):
            res.append(D[d])
            l += len(d)
            for c in d:
                cnt[c] -= 1    
            
    
    res.sort()
    if l <> len(s): 
        print cnt

    return ''.join(res)
        
    
#filespec = "e:/work/code_jam/Getting the Digits.txt"    
filespec = "d:/downloads/A-small-attempt2.in"
#filespec = "d:/downloads/A-large.in"
#filespec = "e:/work/code_jam/test.txt"    
f = open(filespec, 'r')
out = open("e:/work/code_jam/out.txt", 'w')

T = int(f.readline())
for nbr in range(1, T+1):
    s = f.readline().strip()
    
    sol = solve(s)
    s = "Case #%d: %s\n" %(nbr, sol)
    print s,

    out.write(s)
    
f.close()
out.close()
